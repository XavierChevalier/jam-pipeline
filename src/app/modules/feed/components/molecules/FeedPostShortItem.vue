<script setup lang="ts">
  import AppAvatar from '@/app/components/atoms/AppAvatar.vue'
  import { localeFormatDistance } from '@/app/tools/date/Date'
  import { computed, PropType } from 'vue'
  import { Author } from '@/app/modules/feed/models/Author'
  import { isPropertyValid } from '@/app/tools/component-properties/PropertyValidator'
  import { AvailableFeedPostType } from '@/app/modules/feed/models/FeedPostType'
  import { useI18n } from 'vue-i18n'

  const { t } = useI18n()

  const props = defineProps({
    author: {
      type: Object as PropType<Author>,
      required: true,
      validator: isPropertyValid(Author),
    },
    postType: {
      type: String as PropType<keyof typeof AvailableFeedPostType>,
      required: true,
      validator: (value: string) =>
        Object.keys(AvailableFeedPostType).includes(value),
    },
    publicationDate: {
      type: Date,
      required: true,
    },
  })
  const publicationDateDistance = computed(() =>
    localeFormatDistance(props.publicationDate, new Date(), {
      addSuffix: false,
    })
  )
</script>

<template>
  <div class="flex gap-3 items-center">
    <AppAvatar size="sm" :user-name="author.name" :url="author.avatar" />
    <span>{{ t(`postType.${postType}`, { authorName: author.name }) }}</span>
    <div class="flex relative flex-shrink-0 gap-2 items-center ml-auto">
      <div
        class="absolute -left-3 w-0 h-0 rounded-full border-3 border-primary"
      />
      <span class="text-neutral">{{ publicationDateDistance }}</span>
    </div>
  </div>
</template>

<i18n lang="yaml">
en:
  postType:
    userShareSong: '{authorName} shared a song'
    userShareAlbum: '{authorName} shared an album'
    userSharePlaylist: '{authorName} shared a playlist'
    artistReleaseSong: '{authorName} released a new song'
    artistReleaseAlbum: '{authorName} released a new album'
fr:
  postType:
    userShareSong: '{authorName} a partagé un morceau'
    userShareAlbum: '{authorName} a partagé un album'
    userSharePlaylist: '{authorName} a partagé une playlist'
    artistReleaseSong: '{authorName} a sortie un nouveau morceau'
    artistReleaseAlbum: '{authorName} a sortie un nouvel album'
</i18n>
