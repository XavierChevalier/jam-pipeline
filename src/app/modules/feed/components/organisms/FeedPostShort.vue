<script setup lang="ts">
  import FeedPostShortSeeMoreButton from '@/app/modules/feed/components/molecules/FeedPostShortSeeMoreButton.vue'
</script>

<template>
  <div class="flex flex-col gap-3 bg-white rounded p-2">
    <slot />
    <FeedPostShortSeeMoreButton />
  </div>
</template>
