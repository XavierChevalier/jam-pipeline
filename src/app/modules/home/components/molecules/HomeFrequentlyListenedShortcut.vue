<script setup lang="ts">
  import { string } from 'yup'
  import { isPropertyValid } from '@/app/tools/component-properties/PropertyValidator'

  defineProps({
    imageUrl: {
      type: String,
      required: true,
      validator: isPropertyValid(string().required().url()),
    },
  })
</script>

<template>
  <div class="flex items-center bg-white rounded pr-2">
    <img class="rounded" :src="imageUrl" alt="Song or album image" />
    <span class="ml-3">
      <slot />
    </span>
  </div>
</template>
