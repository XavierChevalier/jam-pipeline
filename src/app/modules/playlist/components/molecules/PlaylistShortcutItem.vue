<script setup lang="ts">
  defineProps({
    cover: {
      type: String,
      required: true,
    },
    title: {
      type: String,
      required: true,
    },
    description: {
      type: String,
      required: true,
    },
  })
</script>

<template>
  <div class="flex flex-col min-w-[150px] w-[150px]">
    <img class="rounded" :src="cover" :alt="`Cover of ${title}`" />
    <span class="font-bold mt-2">{{ title }}</span>
    <span class="text-sm">{{ description }}</span>
  </div>
</template>
