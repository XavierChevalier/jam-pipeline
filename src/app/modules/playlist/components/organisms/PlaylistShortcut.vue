<script setup lang="ts">
  import VueHorizontal from 'vue-horizontal'
  import { mdiChevronLeft, mdiChevronRight } from '@mdi/js'
  import AppIcon from '@/app/components/atoms/AppIcon.vue'
  import { useVueHorizontalScroll } from '@/app/composable/VueHorizontalScroll'

  defineProps({
    title: {
      type: String,
      required: true,
    },
  })

  const { horizontalTemplateRef, prev, next, hasPrev, hasNext, onScroll } =
    useVueHorizontalScroll()
</script>

<template>
  <div class="flex flex-col gap-2">
    <header class="flex justify-between items-center">
      <h2 class="text-lg font-bold">{{ title }}</h2>

      <nav class="hidden sm:block">
        <button
          data-test="navigationPrev"
          class="transition-colors"
          :class="{ inactive: !hasPrev }"
          @click="prev"
        >
          <AppIcon :path="mdiChevronLeft" />
        </button>
        <button
          data-test="navigationNext"
          class="transition-colors"
          :class="{ inactive: !hasNext }"
          @click="next"
        >
          <AppIcon :path="mdiChevronRight" />
        </button>
      </nav>
    </header>

    <VueHorizontal
      ref="horizontalTemplateRef"
      :button="false"
      @scroll-debounce="onScroll"
    >
      <slot />
    </VueHorizontal>
  </div>
</template>

<style scoped lang="scss">
  ::v-deep(.v-hl-container) {
    @apply gap-3;
  }

  .inactive {
    @apply text-neutral-light;
  }
</style>
