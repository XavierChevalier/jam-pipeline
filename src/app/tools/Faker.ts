export const generateImageUrl = (size = 50) => `https://picsum.photos/${size}`
