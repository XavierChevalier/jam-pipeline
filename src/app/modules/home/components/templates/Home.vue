<script setup lang="ts">
  import { generateImageUrl } from '@/app/tools/Faker'
  import FeedPostShort from '@/app/modules/feed/components/organisms/FeedPostShort.vue'
  import FeedPostShortItem from '@/app/modules/feed/components/molecules/FeedPostShortItem.vue'
  import HomeFrequentlyListenedShortcuts from '@/app/modules/home/components/organisms/HomeFrequentlyListenedShortcuts.vue'
  import HomeFrequentlyListenedShortcut from '@/app/modules/home/components/molecules/HomeFrequentlyListenedShortcut.vue'
  import PlaylistShortcut from '@/app/modules/playlist/components/organisms/PlaylistShortcut.vue'
  import PlaylistShortcutItem from '@/app/modules/playlist/components/molecules/PlaylistShortcutItem.vue'
</script>

<template>
  <div class="flex flex-col gap-5">
    <FeedPostShort>
      <FeedPostShortItem
        :author="{
          name: 'TheBaronless',
          avatar: generateImageUrl(40),
        }"
        post-type="userShareSong"
        :publication-date="new Date()"
      />
      <FeedPostShortItem
        :author="{
          name: 'TheBaronless',
          avatar: generateImageUrl(40),
        }"
        post-type="userShareAlbum"
        :publication-date="new Date()"
      />
      <FeedPostShortItem
        :author="{
          name: 'TheBaronless',
          avatar: generateImageUrl(40),
        }"
        post-type="userSharePlaylist"
        :publication-date="new Date()"
      />
    </FeedPostShort>

    <HomeFrequentlyListenedShortcuts>
      <HomeFrequentlyListenedShortcut :image-url="generateImageUrl()">
        Transfer
      </HomeFrequentlyListenedShortcut>
      <HomeFrequentlyListenedShortcut :image-url="generateImageUrl()">
        Saas Club
      </HomeFrequentlyListenedShortcut>
      <HomeFrequentlyListenedShortcut :image-url="generateImageUrl()">
        Radar des sorties
      </HomeFrequentlyListenedShortcut>
      <HomeFrequentlyListenedShortcut :image-url="generateImageUrl()">
        Titres likés
      </HomeFrequentlyListenedShortcut>
    </HomeFrequentlyListenedShortcuts>

    <PlaylistShortcut title="Vos mix préférés">
      <PlaylistShortcutItem
        :cover="generateImageUrl(150)"
        title="Polyphia"
        description="New levels new devils"
      />
      <PlaylistShortcutItem
        :cover="generateImageUrl(150)"
        title="Polyphia"
        description="New levels new devils"
      />
      <PlaylistShortcutItem
        :cover="generateImageUrl(150)"
        title="Polyphia"
        description="New levels new devils"
      />
      <PlaylistShortcutItem
        :cover="generateImageUrl(150)"
        title="Polyphia"
        description="New levels new devils"
      />
    </PlaylistShortcut>

    <PlaylistShortcut title="Écoutés récemment">
      <PlaylistShortcutItem
        :cover="generateImageUrl(151)"
        title="Polyphia"
        description="New levels new devils"
      />
      <PlaylistShortcutItem
        :cover="generateImageUrl(151)"
        title="Polyphia"
        description="New levels new devils"
      />
      <PlaylistShortcutItem
        :cover="generateImageUrl(151)"
        title="Polyphia"
        description="New levels new devils"
      />
      <PlaylistShortcutItem
        :cover="generateImageUrl(151)"
        title="Polyphia"
        description="New levels new devils"
      />
    </PlaylistShortcut>
  </div>
</template>
