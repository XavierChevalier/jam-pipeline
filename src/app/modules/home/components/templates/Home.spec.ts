import { StorybookTestsGenerator } from '@/tests/tests-generators/StorybookTestsGenerator'
import * as Stories from './Home.stories'
import FeedPostShort from '@/app/modules/feed/components/organisms/FeedPostShort.vue'
import FeedPostShortItem from '@/app/modules/feed/components/molecules/FeedPostShortItem.vue'
import HomeFrequentlyListenedShortcut from '@/app/modules/home/components/molecules/HomeFrequentlyListenedShortcut.vue'
import HomeFrequentlyListenedShortcuts from '@/app/modules/home/components/organisms/HomeFrequentlyListenedShortcuts.vue'
import PlaylistShortcut from '@/app/modules/playlist/components/organisms/PlaylistShortcut.vue'
import PlaylistShortcutItem from '@/app/modules/playlist/components/molecules/PlaylistShortcutItem.vue'
import { mountWithPlugins } from '@/tests/tools/MountWithPlugins'

describe('Home', () => {
  const component = StorybookTestsGenerator.fromStoriesExports(Stories)
    .snapshotEachStories()
    .getComponent()

  const wrapper = mountWithPlugins(component)
  it('should render feed post shorts', () => {
    wrapper.getComponent(FeedPostShort)
    expect(
      wrapper.findAllComponents(FeedPostShortItem).length
    ).toBeGreaterThanOrEqual(1)
  })

  it('should render frequently listened shortcuts', () => {
    wrapper.getComponent(HomeFrequentlyListenedShortcuts)
    expect(
      wrapper.findAllComponents(HomeFrequentlyListenedShortcut).length
    ).toBeGreaterThanOrEqual(1)
  })

  it('should render playlists shortcuts', () => {
    const playlistShortcuts = wrapper.findAllComponents(PlaylistShortcut)
    expect(playlistShortcuts.length).toBeGreaterThanOrEqual(1)
    playlistShortcuts.forEach((playlistShortcut) => {
      expect(
        playlistShortcut.findAllComponents(PlaylistShortcutItem).length
      ).toBeGreaterThanOrEqual(1)
    })
  })
})
