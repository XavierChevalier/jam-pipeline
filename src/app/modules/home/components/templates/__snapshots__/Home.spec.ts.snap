// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home stories snapshots [snapshot] should render the "Default" component story correctly 1`] = `
<div
  class="flex flex-col gap-5"
>
  <anonymous-stub />
  <anonymous-stub />
  <anonymous-stub
    title="Vos mix préférés"
  />
  <anonymous-stub
    title="Écoutés récemment"
  />
</div>
`;
